lgc/patch/MeshTaskShader.cpp

```diff
@@ -820,6 +844,152 @@ void MeshTaskShader::writeTaskPayload(Value *writeValue, Value *byteOffset) {
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_or;
+    break;
+  case AtomicRMWInst::Xor:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_xor;
+    break;
+  case AtomicRMWInst::Max:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_smax;
+    break;
+  case AtomicRMWInst::Min:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_smin;
+    break;
+  case AtomicRMWInst::UMax:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_umax;
+    break;
+  case AtomicRMWInst::UMin:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_umin;
+    break;
+  case AtomicRMWInst::FAdd:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_fadd;
+    break;
+  case AtomicRMWInst::FMax:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_fmax;
+    break;
+  case AtomicRMWInst::FMin:
+    intrinsic = Intrinsic::amdgcn_raw_buffer_atomic_fmin;
+    break;
+  default:
+    llvm_unreachable("Unexpected atomic operation!");
+    break;
+  }
+
+  if (ordering == AtomicOrdering::Release || ordering == AtomicOrdering::AcquireRelease ||
+      ordering == AtomicOrdering::SequentiallyConsistent)
+    m_builder->CreateFence(AtomicOrdering::Release, SyncScope::System);
+
+  Value *atomicCall = m_builder->CreateIntrinsic(
+      intrinsic, inputValue->getType(),
+      {inputValue, payloadRingBufDesc, byteOffset, payloadRingEntryOffset, m_builder->getInt32(0)});
+
+  if (doubleToInt64)
+    atomicCall = m_builder->CreateBitCast(atomicCall, m_builder->getDoubleTy());
+
+  if (ordering == AtomicOrdering::Release || ordering == AtomicOrdering::AcquireRelease ||
+      ordering == AtomicOrdering::SequentiallyConsistent)
+    m_builder->CreateFence(AtomicOrdering::Acquire, SyncScope::System);
+
+  return atomicCall;
+}
+
+// =====================================================================================================================
+// Create a task payload atomic compare-and-swap.
+//
+// @param ordering : Atomic ordering
+// @param inputValue : Input value
+// @param comparatorValue : Value to compare against
+// @param byteOffset : Byte offset within the payload structure
+// @returns : Original value read from the task payload
+Value *MeshTaskShader::taskPayloadAtomicCompareSwap(AtomicOrdering ordering, Value *inputValue, Value *comparatorValue,
+                                                    Value *byteOffset) {
+  auto entryPoint = m_builder->GetInsertBlock()->getParent();
+  assert(getShaderStage(entryPoint) == ShaderStageTask);
+
+  assert(inputValue->getType()->isIntegerTy() || inputValue->getType()->isFloatingPointTy());
+
+  auto payloadRingBufDesc = m_pipelineSysValues.get(entryPoint)->getTaskPayloadRingBufDesc();
+  auto payloadRingEntryOffset = getPayloadRingEntryOffset(entryPoint);
+
+  if (inputValue->getType()->isIntegerTy(64)) {
+    // NOTE: HW doesn't have buffer_atomic_cmpswap_x2 instruction, we resort to global_atomic_cmpswap_x2.
+
+    // 48-bit GPU address of from the buffer descriptor: dword1[15:0] + dword0
```

Comment:
Do we have this workaround code somewhere else already?

Reply:
Yes, this workaround is in PatchBufferOp.cpp when we handle normal shader storage buffer. See this: `PatchBufferOp::visitAtomicCmpXchgInst`.

Reply:
I see. I suppose it's not a big deal to leave it duplicated for now.